cmd/observer/node_builder/observer_builder.go

```diff
@@ -0,0 +1,865 @@
+
+			builder.initMiddleware(observerNodeID, node.Metrics.Network, libP2PFactory, msgValidators...)
+
+			return nil
+		})
+}
+
+// Build enqueues the sync engine and the follower engine for the observer.
+// Currently, the observer only runs the follower engine.
+func (builder *ObserverServiceBuilder) Build() (cmd.Node, error) {
+	builder.BuildConsensusFollower()
+	return builder.FlowNodeBuilder.Build()
+}
+
+// enqueuePublicNetworkInit enqueues the observer network component initialized for the observer
+func (builder *ObserverServiceBuilder) enqueuePublicNetworkInit() {
+
+	builder.Component("unstaked network", func(node *cmd.NodeConfig) (module.ReadyDoneAware, error) {
+		var heroCacheCollector module.HeroCacheMetrics = metrics.NewNoopCollector()
+		if builder.HeroCacheMetricsEnable {
+			heroCacheCollector = metrics.NetworkReceiveCacheMetricsFactory(builder.MetricsRegisterer)
+		}
+		receiveCache := netcache.NewHeroReceiveCache(builder.NetworkReceivedMessageCacheSize,
+			builder.Logger,
+			heroCacheCollector)
+
+		err := node.Metrics.Mempool.Register(metrics.ResourceNetworkingReceiveCache, receiveCache.Size)
+		if err != nil {
+			return nil, fmt.Errorf("could not register networking receive cache metric: %w", err)
+		}
+
+		// topology is nil since it is automatically managed by libp2p
+		net, err := builder.initNetwork(builder.Me, builder.Metrics.Network, builder.Middleware, nil, receiveCache)
+		if err != nil {
+			return nil, err
+		}
+
+		builder.Network = converter.NewNetwork(net, engine.SyncCommittee, engine.PublicSyncCommittee)
+
+		builder.Logger.Info().Msgf("network will run on address: %s", builder.BindAddr)
+
+		idEvents := gadgets.NewIdentityDeltas(builder.Middleware.UpdateNodeAddresses)
+		builder.ProtocolEvents.AddConsumer(idEvents)
+
+		return builder.Network, nil
+	})
+}
+
+// enqueueConnectWithStakedAN enqueues the upstream connector component which connects the libp2p host of the observer
+// AN with the staked AN.
+// Currently, there is an issue with LibP2P stopping advertisements of subscribed topics if no peers are connected
+// (https://github.com/libp2p/go-libp2p-pubsub/issues/442). This means that an observer could end up not being
+// discovered by other observers if it subscribes to a topic before connecting to the staked AN. Hence, the need
+// of an explicit connect to the staked AN before the node attempts to subscribe to topics.
+func (builder *ObserverServiceBuilder) enqueueConnectWithStakedAN() {
+	builder.Component("upstream connector", func(_ *cmd.NodeConfig) (module.ReadyDoneAware, error) {
+		return consensus_follower.NewUpstreamConnector(builder.bootstrapIdentities, builder.LibP2PNode, builder.Logger), nil
+	})
+}
+
+func (builder *ObserverServiceBuilder) attachRPCEngine() {
+	builder.Component("RPC engine", func(node *cmd.NodeConfig) (module.ReadyDoneAware, error) {
+                proxy, err := apiservice.NewFlowAPIService(builder.bootstrapIdentities, builder.PeerUpdateInterval)
```

Comment:
Makes sense.

Reply:
access-client-timeout Fixed: https://github.com/GetElastech/flow-go/commit/f032844f3f43742c738633374d935961752ab4f8
I used a different variable name provided that we are in observer.

Reply:
Quick question: "builder.bootstrapIdentities is going to include Addresss with the upstream node's public libp2p port. Rather than using these directly, maybe we should add a new CLI flag that allows specifying the host:port for the upstream ANs gRPC address"

Does this mean that we can point to access gRPC or access libp2p. How should I specify the public key for the API? This may require a separate GH issue.

Reply:
> Does this mean that we can point to access gRPC or access libp2p.

No, it must connect to a gRPC interface on the upstream node.

> How should I specify the public key for the API?

I think it's ok to only support plaintext gRPC for now, and we can add secured gRPC later.

> This may require a separate GH issue.

That's fine, but keep in mind the proxy won't work until this is fixed.

Reply:
Fixed: https://github.com/GetElastech/flow-go/commit/31ea1e59664f9330bb4140de2d72acc121779eaa